File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/UserJoined.java
package org.emulinker.kaillera.controller.v086.protocol;

import com.google.auto.value.AutoValue;
import com.google.common.base.Strings;
import java.nio.ByteBuffer;
import org.emulinker.kaillera.controller.messaging.*;
import org.emulinker.kaillera.model.KailleraUser;
import org.emulinker.kaillera.pico.AppModule;
import org.emulinker.util.*;

@AutoValue
public abstract class UserJoined extends V086Message {
  public static final byte ID = 0x02;
  private static final String DESC = "User Joined";

  public abstract String username();

  public abstract int userId();

  public abstract long ping();

  public abstract byte connectionType();

  public static AutoValue_UserJoined create(
      int messageNumber, String username, int userId, long ping, byte connectionType)
      throws MessageFormatException {
    V086Message.validateMessageNumber(messageNumber, DESC);

    if (Strings.isNullOrEmpty(username))
      throw new MessageFormatException(
          "Invalid " + DESC + " format: userName.length == 0, (userID = " + userId + ")");

    if (userId < 0 || userId > 65535)
      throw new MessageFormatException(
          "Invalid " + DESC + " format: userID out of acceptable range: " + userId);

    if (ping < 0 || ping > 2048) // what should max ping be?
    throw new MessageFormatException(
          "Invalid " + DESC + " format: ping out of acceptable range: " + ping);

    if (connectionType < 1 || connectionType > 6)
      throw new MessageFormatException(
          "Invalid " + DESC + " format: connectionType out of acceptable range: " + connectionType);

    return new AutoValue_UserJoined(
        messageNumber, ID, DESC, username, userId, ping, connectionType);
  }

  // TODO(nue): Get rid of this.
  @Override
  public String toString() {
    return getInfoString()
        + "[userName="
        + username()
        + " userID="
        + userId()
        + " ping="
        + ping()
        + " connectionType="
        + KailleraUser.CONNECTION_TYPE_NAMES[connectionType()]
        + "]";
  }

  @Override
  public int getBodyLength() {
    return getNumBytes(username()) + 8;
  }

  @Override
  public void writeBodyTo(ByteBuffer buffer) {
    EmuUtil.writeString(buffer, username(), 0x00, AppModule.charsetDoNotUse);
    UnsignedUtil.putUnsignedShort(buffer, userId());
    UnsignedUtil.putUnsignedInt(buffer, ping());
    buffer.put(connectionType());
  }

  public static UserJoined parse(int messageNumber, ByteBuffer buffer)
      throws ParseException, MessageFormatException {
    if (buffer.remaining() < 9) throw new ParseException("Failed byte count validation!");

    String userName = EmuUtil.readString(buffer, 0x00, AppModule.charsetDoNotUse);

    if (buffer.remaining() < 7) throw new ParseException("Failed byte count validation!");

    int userID = UnsignedUtil.getUnsignedShort(buffer);
    long ping = UnsignedUtil.getUnsignedInt(buffer);
    byte connectionType = buffer.get();

    return UserJoined.create(messageNumber, userName, userID, ping, connectionType);
  }
}

File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/UserJoined.kt
package org.emulinker.kaillera.controller.v086.protocol

import java.nio.ByteBuffer
import org.emulinker.kaillera.controller.messaging.MessageFormatException
import org.emulinker.kaillera.controller.messaging.ParseException
import org.emulinker.kaillera.controller.v086.protocol.V086Message.Companion.validateMessageNumber
import org.emulinker.kaillera.model.ConnectionType
import org.emulinker.kaillera.pico.AppModule
import org.emulinker.util.EmuUtil
import org.emulinker.util.UnsignedUtil

data class UserJoined
    @Throws(MessageFormatException::class)
    constructor(
        override val messageNumber: Int,
        val username: String,
        val userId: Int,
        val ping: Long,
        val connectionType: ConnectionType
    ) : V086Message() {

  override val messageId = ID

  init {
    validateMessageNumber(messageNumber)
    if (username.isBlank()) throw MessageFormatException("Empty username: $username")
    require(userId in 0..65535) { "UserID out of acceptable range: $userId" }
    require(ping in 0..2048) { "Ping out of acceptable range: $ping" }
  }

  override val bodyLength = getNumBytes(username) + 8

  public override fun writeBodyTo(buffer: ByteBuffer) {
    EmuUtil.writeString(buffer, username, 0x00, AppModule.charsetDoNotUse)
    UnsignedUtil.putUnsignedShort(buffer, userId)
    UnsignedUtil.putUnsignedInt(buffer, ping)
    buffer.put(connectionType.byteValue)
  }

  companion object {
    const val ID: Byte = 0x02

    @Throws(ParseException::class, MessageFormatException::class)
    fun parse(messageNumber: Int, buffer: ByteBuffer): UserJoined {
      if (buffer.remaining() < 9) throw ParseException("Failed byte count validation!")
      val userName = EmuUtil.readString(buffer, 0x00, AppModule.charsetDoNotUse)
      if (buffer.remaining() < 7) throw ParseException("Failed byte count validation!")
      val userID = UnsignedUtil.getUnsignedShort(buffer)
      val ping = UnsignedUtil.getUnsignedInt(buffer)
      val connectionType = buffer.get()
      return UserJoined(
          messageNumber, userName, userID, ping, ConnectionType.fromByteValue(connectionType))
    }
  }
}
